lib/compat/wordpress-6.3/template-parts-preloading.php

```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * Patches resources loaded by the block editor page
+ * to include Template Part posts.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Preloads requests needed for common Template Part posts
+ *
+ * @param array $preload_paths    Preload paths to be filtered.
+ * @return array
+ */
+function gutenberg_preload_template_parts( $preload_paths, $context ) {
+
+	// Limit to the Site Editor.
+	if ( ! empty( $context->name ) && 'core/edit-site' !== $context->name ) {
+		return $preload_paths;
+	}
+
+	$theme_slug = wp_get_theme()->get_stylesheet();
+
+	$template_parts_rest_route = rest_get_route_for_post_type_items(
+		'wp_template_part'
+	);
+
+	$standard_template_parts = array( 'header', 'footer' );
```

Comment:
This asserts that the template part names will always be the same. Given that those can often change and are dynamic, I think this optimization won't help us much. 

Furthermore, this asserts that these are the most widely used template parts, which often won't be the case. I've seen some themes with multiple headers and footers (usually one for the home page, one for the inner pages), and this may not only catch any of them; but also incorrectly assume that these are the most important ones. In the scenario of "one header/footer pair for home page, one pair for all other pages", maybe those represent the ones used for the home page and will end up being preloaded regardless that the other header/footer pair is used more often. 

I wonder if we could try to be smarter about those and preload template parts that are actually widely used across the current templates.

Reply:
> This asserts that the template part names will always be the same. Given that those can often change and are dynamic, I think this optimization won't help us much.

Yes this is entirely true. My rationale is that the majority of Themes will likely use `header` and `footer` for their Template Part names for those locations. TBH we could ditch footer and just load `header.

> I've seen some themes...

I think this is the key statement. Optimise for the majority but don't break it for the minority. Will this make the experience worse for those edge case themes and does that outweigh the benefit for the majority?

> I wonder if we could try to be smarter about those and preload template parts that are actually widely used across the current templates.

That would require another request for templates. It starts to become self defeating.

Reply:
This has been fixed by only loading template parts defined in `theme.json`.